### ISSUE TITLE
[AppCenter Data] Read documents with custom PartitionKey

### ISSUE THREAD

**Issue by devStarscream** [microsoft/appcenter#681]
**Describe the solution you'd like**
User `AppCenter.Data` for accessing documents where "PartitionKey" value is different either from "user" or "readonly".
Since `Data.ReadAsync<T>(string id, string partition)` takes a `string` value for the partition, it means that I could provide my custom values.

**Describe alternatives you've considered**
Current alternative is to access CosmosDB directly, avoiding the `AppCenter.Data` SDK.
Maybe I'm wrong, but there is no meaning on excluding documents with dedicated values for `PartitionKey`.

**Additional context**
When I try to read documents with custom `PartitionKey` value

`await Data.ReadAsync<ToDoItem>("document-id", "A CUSTOM VALUE");`

I get the following error: 

`com.microsoft.appcenter.data.exception.DataException: Partition name can be either 'readonly' or 'user' but not 'A CUSTOM VALUE'`.

**Comment by Zakeelm** [comment]
The way our current architecture is set up, collections must use the value `PartitionKey`, which in turn means we don't support custom PartitionKey values. The current reason we do this is because it enables us to differentiate between private and public documents. The String `PartitionKey` parameter should either be `DefaultPartitions.UserDocuments` or `DefaultPartitions.AppDocuments`. 

This is something we can investigate though. Can you tell me more on why custom values are important in your use case and what limitations the current architecture has to you due to lack of custom value capabilities?

**Comment by devStarscream** [comment]
Thank you for your answer @Zakeelm.

I don't know which are you architecural desing criteria, but in my case I would like to _partition_ my data based on the `user-id` without being worried about restrictions: public vs private.

According to [the documentation](https://docs.microsoft.com/en-us/appcenter/data/#how-it-works)) 

> The list of all the top restaurants in your area would be visible to all app users. Say you define a top restaurant using ratings and popularity metrics. As this information isn't tied to a specific user and is global, **it would be stored in a public document and exist in the public read only logical partition**.

and to [this other documentation ](https://docs.microsoft.com/en-us/appcenter/sdk/data/xamarin#creating-a-document)

> The partition that the document will live in. You will most likely be using the `DefaultPartitions.UserDocuments` option to **store this within a specific user's partition**. 
>Note
>
>**Public documents are read-only.**

I can use `DefaultPartitions.AppDocuments` (i.e., `readonly`) to put all my data into the `public` partition, which is accesible to every one.
It means that there is one and only one `readonly` partition, which will grow very very fast and it will be the biggest partition in my collection (risking also to exceed the limit of 10GB per partition).

According to CosmosDB's statements:
> Any field / value can be used for partioning data

In my opinion, it would be easier to allow setting a custom value for the `PartitionKey` field, without being worried about the _privateness_ and the _publicness_ of the data.
This solution increases the DoF (Degrees of Freedom) on structuring data, also at AppCenter.Data SDK level.

For data _publicness_ / _privateness_ the current approach of concatenation of `user-` string can still be used. While `readonly` and `other strings` can be allowed for further partitions.

Currently I have removed `AppCenter.Data` SDK from my solution (iOS and Anroid apps), and based everything on `DocumentClient`.
Nevertheless, I think that my current solution is not very smart, considering that `AppCenter.Data` SDK has been "invented" for keeping Azure intergration easier.

**Comment by Zakeelm** [comment]
From our perspective, we believed 10GB per partition was sizable for the average documents. Our recommendation when dealing with potential huge documents is putting the large payload in blob storage and specifying a link to it within the Cosmos DB document. 

We also plan on implementing shared documents, which will likely belong in a shared partition. I think this may give you the desired DoF you would like as it can be as "private" or "pubic" as you would like and gives you much more freedom in how we you structure your data. 

I do understand completely where you're coming from with the variability of partitions. While we don't have current plans to change our current architecture, I can definitely leave this open for now, and give others the opportunity to give feedback on this as well. If this is something others are interested in, we can definitely investigate this further internally.

**Comment by baskren** [comment]
@Zakeelm - **Shared Documents** would address the gap I see in AppCenter.Data for my applications.  However, I did not see this on the [Roadmap](https://github.com/Microsoft/appcenter/wiki/Roadmap).  Is **Share Documents** a planned feature or just a feature under consideration?

**Comment by baskren** [comment]
@Zakeelm - nevermind.  I just found #864

**Comment by Zakeelm** [comment]
@baskren Yep! Shared Documents is planned work, would the scenarios outlined in [this comment](https://github.com/microsoft/appcenter/issues/864#issuecomment-521060202) be enough for your use cases or are there additional scenarios you would like to see within shared documents?

**Comment by Zakeelm** [comment]
Thank you so much for all of your requests. We announced this week in a [blog post]( https://devblogs.microsoft.com/appcenter/app-center-mbaas-retirement/) that we’re retiring the Auth and Data services later this year and recommending customers transition to Azure Active Directory B2C and Azure Cosmos DB instead. Let me know if you have any additional questions about this transition and i'll make sure to get them answered! Closing this issue.
